src/concurrency/sync.rs

```diff
@@ -149,18 +158,99 @@ struct FutexWaiter {
     bitset: u32,
 }
 
+declare_id!(InitOnceId);
+
+struct InitOnceWaiter<'mir, 'tcx> {
+    thread: ThreadId,
+    callback: Box<dyn MachineCallback<'mir, 'tcx> + 'tcx>,
+}
```

Comment:
Please add doc comments. In particular I am wondering what the callback is doing.